### ISSUE TITLE
[Bug]: MLX backend ignores saved adapter checkpoints on --resume and restarts training from scratch

### ISSUE THREAD

**Issue by imahsanali** [MakazhanAlpamys/Soup#634]
### Problem

When running SFT fine-tuning with backend: mlx on Apple Silicon, training cannot be resumed after an interruption (or Metal crash) using soup train --config soup.yaml --resume auto or --resume <output_dir>.

Document attached of fix i applied:

[MLX_RESUME_FIX_EXPLANATION.md](https://github.com/user-attachments/files/31698235/MLX_RESUME_FIX_EXPLANATION.md)

### Proposed solution

_resolve_checkpoint() in train.py only scans for HuggingFace/PyTorch directories (checkpoint-*), failing to detect MLX *_adapters.safetensors files.
MLXSFTTrainerWrapper.train() in mlx_sft.py hardcodes a fallback ("MLX backend does not support --resume yet; starting training from scratch.") and completely ignores resume_from_checkpoint.
The dataset iterator in mlx_lm does not offset/skip already processed batches to the saved iteration index (start_step).
Steps to Reproduce:

Start MLX training: soup train --config soup.yaml (saves checkpoints e.g., 0011800_adapters.safetensors).
Stop or let training crash/interrupt after step 100+.
Run soup train --config soup.yaml --resume auto (or --resume output_dir).
Result: Soup either reports No checkpoint found to resume from or restarts from iteration 1 instead of resuming from the saved checkpoint.
Expected Behavior: soup train --resume auto should detect the latest saved *_adapters.safetensors, load the adapter weights via model.load_weights(), skip already processed training batches, and continue training seamlessly from step N + 1.

Document for fix is attached below

### Area

Training (SFT/DPO/GRPO)

**Comment by MakazhanAlpamys** [comment]
Thank you — this is a good report, and both mechanisms check out against the source. Confirmed on `main` at `74ac1e3`:

- `src/soup_cli/commands/train.py:2007` — `_resolve_checkpoint("auto", ...)` filters to `d.is_dir() and d.name.startswith("checkpoint-")`. An MLX run writes `NNNNNNN_adapters.safetensors` **files**, so the list is empty and `--resume auto` returns `None`. That is your "No checkpoint found to resume from".
- `src/soup_cli/trainer/mlx_sft.py:181-185` — `train()` prints `MLX backend does not support --resume yet; starting training from scratch.` and then `del`s nothing, uses nothing: `resume_from_checkpoint` is never read again.

To be precise about what I did and did not verify: I read the code, I did not run it. I have no Apple Silicon box, so I have not reproduced the crash-and-resume cycle end to end, and I will not claim I have. Everything above is a source-level confirmation of the two mechanisms you named.

**Two remarks that will matter if you open the PR.**

First, the warning at `mlx_sft.py:183` is at least honest — it says it is starting from scratch. That puts this a tier below the failure mode this project treats as worst (silent). It is still a real bug: the flag is accepted, documented, and does nothing, which is the same "validated and read by nothing" pattern that ran through v0.73.3.

Second, your third item — *"the dataset iterator does not offset to the saved iteration index"* — is the hard one and I would like it separated from the first two in the PR, or scoped out of it. Detecting the adapter file and calling `model.load_weights()` is mechanical. Skipping N already-processed batches is a **resume-determinism** claim: it is only correct if the iteration order is reproducible from the seed, and `training.seed` / `training.data_seed` are wired into the SFT trainer only (see #353). If MLX's `create_dataset` / `CacheDataset` path does not consume that seed, then "skip 100 batches" skips 100 batches of a *different* permutation and quietly trains on a distribution you did not intend — which would be worse than restarting. If you have measured it, say what you measured; if not, land loading + detection first and file the offset separately.

**A PR is very welcome, and the work is yours** — I would rather have your fix than write my own from your document. What I ask for on any PR here:

1. Tests that fail before the change. For the detection half that is pure and CPU-testable: `_resolve_checkpoint` against a temp dir containing `0011800_adapters.safetensors` needs no MLX at all, so most of this is verifiable on any machine.
2. A mutation check in the PR body — revert the core of your fix, show which named test fails. If none does, the test is not pinning anything.
3. A `changelog.d/` fragment, and the resume support stated in the docs where the MLX backend is described.
4. If part of it can only be verified on Apple Silicon, say so plainly rather than implying a run you did not do. That is the standard I hold myself to above and it is not a higher bar for contributors.

One design question I would rather settle before you build: `_resolve_checkpoint` currently returns a directory for every backend. Making it return a file path for MLX changes its contract for every caller. I would prefer a backend-aware resolver — the MLX branch resolves MLX artifacts — over widening the shared one, but if you have a reason to prefer the other shape, argue it in the PR and I will read it.

Also: your attached `MLX_RESUME_FIX_EXPLANATION.md` is a GitHub file attachment. Please paste the substance into the PR description or a comment as well — attachments are easy to lose and are not searchable from the issue.

Labelled `bug` + `help wanted`. I tried to assign it to you and GitHub refused — the API only accepts assignees with write access, which you do not have and do not need. Treat it as yours: I will not take it without saying so here first, and if you would rather I did, just say.

**Comment by MakazhanAlpamys** [comment]
@imahsanali — a heads-up you are owed, and a question.

Another contributor has opened **#639** implementing this fix. They did not see our exchange (they did not comment here first, which is the project's claiming convention and is on them to follow, not on you to police), and I am not merging it yet.

I told you above:

> Treat it as yours: I will not take it without saying so here first, and if you would rather I did, just say.

That still holds. So:

**Do you want to open your fix?** If yes, yours goes first and I review it on its merits. Take the week — there is no race here, and "life got in the way" needs no explanation. If you would rather not, say so in one line and #639 becomes the fix with no hard feelings in either direction.

There is a third option worth naming, because it may be the best outcome: if both fixes exist, I will say openly which half of each is stronger and ask whether you want to combine them, rather than quietly picking one and leaving the other author to work out why. #639 carries a 223-line test file; your write-up described the runtime behaviour in more detail than the tests would. Those are complementary, not competing.

**If you do open it, the two things I asked for still stand**, and the second is the one that matters most here:

1. Tests that fail before the change. The detection half needs no MLX at all — `_resolve_checkpoint` against a temp directory containing `0011800_adapters.safetensors` is pure and runs on any machine.
2. Keep the **batch-offset** item separate from checkpoint detection, or scope it out. Detecting the adapter and calling `model.load_weights()` is mechanical; skipping N already-processed batches is a *resume-determinism* claim that is only correct if the iteration order reproduces from the seed — and `training.seed` reaches the SFT trainer only (#353). If MLX's `create_dataset` / `CacheDataset` path does not consume that seed, "skip 100 batches" skips 100 batches of a **different permutation**, which is worse than restarting because it is silent.

No deadline pressure beyond the week, and tell me if a week is not enough — I would rather wait than have your report turn into someone else's patch.

**Comment by Megadurck** [comment]
I’d like to take this issue. I’ve identified the root cause in the resume checkpoint detection logic and am preparing a fix with a regression test.

**Comment by imahsanali** [comment]
Hi @MakazhanAlpamys, thanks for the opportunity! Yes, I’d be happy to work on the fix and submit a PR. I’ll need about a week to make the changes and ensure they meet the repository’s requirements.

**Comment by MakazhanAlpamys** [comment]
Update, and it strengthens the case for your input rather than replacing it.

I have reviewed #639 and **requested changes on technical grounds** — separate from the fairness hold, and I said so there explicitly so the author knows which is which. Two of the findings are precisely where someone who has actually run this path would be faster than either of us reading code:

1. **`experiment_name` is still broken.** The PR's MLX resolver copied the transformers nesting (`base / experiment_name`), but MLX writes flat — `trainer/mlx_sft.py:195` is `output_dir = Path(cfg.output)`, while `trainer/sft.py:752-753` nests. I verified both sides on `main`. So any `soup.yaml` carrying an `experiment_name` still gets `No checkpoint found to resume from.` — your exact symptom, unfixed.

2. **`load_weights(..., strict=False)` with no post-load check.** This repo already recorded, in that same file, that `strict=False` can drop every LoRA tensor in silence and leave an adapter whose generations are bit-identical to the base model (#392). If a checkpoint was written under a different `lora.r` or different `target_modules`, the user would see two green messages over a run that warm-started from nothing. I could not test this — no Apple Silicon here.

**Question for you, if you have a moment even without opening a PR:** when your fix loaded a checkpoint, did you verify anything actually landed — a parameter count, a loss that continued rather than restarting, generations that differed from the base? That single observation would settle finding 2 for both of us. Answering it in a comment costs you nothing and is genuinely useful whether or not you take the PR.

The week I offered still stands, unchanged.

**Comment by imahsanali** [comment]
**Finding 2 first, since I have hard numbers:**

I ran the exact check you described — parameter count and value comparison before and after `load_weights`:

```text
Trainable param count BEFORE load: 3,686,400
Trainable param count AFTER load:  3,686,400

Counts match (same LoRA shape): True

model.layers.0.self_attn.q_proj.lora_a
  BEFORE: 0.002468
  AFTER: -0.014009

model.layers.0.self_attn.q_proj.lora_b
  BEFORE: 0.0
  AFTER: -0.003909

Weights actually changed after load_weights: True
```

The weights **did land**. The LoRA layer count is identical (3,686,400 trainable parameters, matching `r=16` across Q/V projections on all layers of Qwen2.5-Coder-3B), and the values differ non-trivially from freshly-initialised weights.

This specific config (`r=16`, `target_modules: auto` → `q_proj` + `v_proj`) is internally consistent, so `strict=False` does not silently drop anything here. **Your concern is valid as a general guard for mismatched configs, but it is not what happened in this run.**

---

**Finding 1 — `experiment_name` nesting — you are correct:**

My `_resolve_checkpoint` fix copied the `base / experiment_name` nesting pattern from the transformers-side resolver, but `mlx_sft.py` line 189 does `output_dir = Path(cfg.output)` with no `experiment_name` sub-folder.

So on any `soup.yaml` that sets `experiment_name`, `auto` still returns `None`. This is a genuine unfixed bug in the PR. My test config has no `experiment_name` set, which is why I didn't hit it.

The correct fix for `_resolve_checkpoint` on the MLX path should **not** append `experiment_name` before scanning for `*_adapters.safetensors`:

```python
# MLX writes flat into cfg.output — do NOT nest experiment_name here

mlx_base = Path(output_dir)

mlx_checkpoints = sorted(
    [
        f for f in mlx_base.iterdir()
        if f.is_file()
        and f.name.endswith("_adapters.safetensors")
        and f.name.split("_")[0].isdigit()
    ],
    key=lambda f: int(f.name.split("_")[0]),
)
```

The `experiment_name` nesting should only apply to the HuggingFace `checkpoint-*` branch.

---

**Summary:**

* Finding 1 (`experiment_name`): ✅ **confirmed broken, your analysis is correct**
* Finding 2 (`strict=False` silent drop): ✅ **weights confirmed to land correctly for matching configs** (but your guard suggestion — checking that at least N LoRA tensors were actually updated — is worth adding as a defensive check regardless)

**Comment by MakazhanAlpamys** [comment]
Thank you — that is the measurement I could not make from reading code, and it changes one of my two findings rather than confirming both. Recording that plainly.

**Finding 2 is narrowed, and the narrowing is yours, not mine.** I wrote that `load_weights(strict=False)` *could* drop tensors silently; you showed that in your run it did not — 3,686,400 trainable parameters before and after, `lora_b` moving off its zero init, values non-trivially different from a fresh initialisation. `lora_b` starting at exactly 0.0 and ending at -0.003909 is the strongest single number there, because a freshly-initialised `lora_b` is zero by construction, so a value that is not zero cannot have come from anywhere but the file. My concern was a hypothesis about a failure mode; you turned it into a bounded statement: **for a config whose shape matches the checkpoint, the weights land.** What remains is a guard for the mismatched case, which is a smaller and different claim, and I will treat it that way rather than as a defect.

I have said the same on #639 so the author is not held against a finding that has been measured down.

**Finding 1 stands and you confirmed the mechanism independently**, including the reason it was invisible: your test config sets no `experiment_name`, so the nesting bug cannot reproduce on it. That is worth more than the fix itself — it explains why the PR passed its own tests, and it is the shape of thing that a test written from the reproduction alone will keep missing.

Two notes on your proposed patch, both about making the fix fail loudly rather than quietly:

1. **The flat-vs-nested split should be derived, not duplicated.** `trainer/mlx_sft.py` writes flat (`output_dir = Path(cfg.output)`); `trainer/sft.py` nests (`base / experiment_name`). If the resolver hardcodes "MLX is flat" as a second copy of that knowledge, the two can drift the next time either writer changes, and the symptom is again a silent `No checkpoint found to resume from.` Prefer having the resolver ask the same helper the writer uses, or — if that is too large — a test that fails if the writer's layout stops matching the resolver's assumption.
2. **`f.name.split("_")[0].isdigit()` accepts a name like `12_extra_adapters.safetensors`** and orders it by `12`. Probably harmless, but worth deciding deliberately rather than inheriting from the glob.

On timing: take your week, that is fine, and the hold on #639 runs to roughly 2026-09-08 as promised. If you would rather not carry the whole fix, the `experiment_name` half alone is a complete, mergeable contribution and I will credit it as such.

**Comment by MakazhanAlpamys** [comment]
@imahsanali — an update you are owed, and it involves me correcting myself rather than you.

**I have lifted the hold I placed on #639.** When I offered you this work on 2026-09-01 at 20:58, @SID-6921's PR had already been open for 28 minutes. I did not check, and I then paused their PR in favour of an offer I made after it existed. Two days later I hit the same collision on #644/#646 and ruled the opposite way — protecting the PR that came first. One rule for one contributor and its inverse for another is not a rule, and #639 is the one that lost five days to it.

So #639 proceeds on its merits. It is not merged: I have just requested two small test additions.

**What you contributed is not diminished by that, and I want it stated plainly rather than softened.** You ran the measurement I could not — parameter count identical before and after `load_weights`, `lora_b` moving from exactly 0.0 to -0.003909 — and it **retracted one of my two blocking findings**. A fresh `lora_b` is zero by construction, so a non-zero value cannot have come from anywhere but the file; that is the kind of evidence that closes a question instead of continuing it. You also confirmed the `experiment_name` mechanism independently and explained why your own reproduction could not hit it, which is more useful than the fix.

#639's changelog fragment credits you by name (`reported and diagnosed by @imahsanali`) and I have asked its author to carry that into the PR body as well, so it appears on the merge commit. That credit stands regardless of whose patch lands.

**If you still want a first patch here, I would rather find you a good one than have you race a PR that is nearly finished.** Three that are genuinely open, self-contained, and yours if you want them:

- **#650** — the training panel reports a trough instead of a peak, and the *label* is the real defect. Currently offered to its own reporter until tomorrow, so ask and I will tell you if it frees up.
- **#623 (remaining half)** — a CPU-only assertion that `_build_source(..., pin=False, tier="ram")` yields `.pinned is False`. Needs no GPU, and closes a gap that currently survives on all nine CI cells.
- **#649** — the batch probe approves sizes that spill to host memory under WDDM. Harder, and it needs the reporter's platform to verify live, but the decision logic is testable anywhere.

Say the word on any of them and I will assign it before anyone else picks it up — which is the step I skipped here.

**Comment by MakazhanAlpamys** [comment]
Fixed in #639 by @SID-6921, merged as `7682bc2`.

`--resume` now recognises mlx-lm's flat `NNNNNNN_adapters.safetensors` snapshots, picks the highest step by numeric value, and loads the checkpoint **after** `_apply_lora` — loading into a model with no adapter yet writes nowhere, which was the second half of the bug.

**Your `experiment_name` finding is what made the fix correct.** The first version copied the transformers `base / experiment_name` nesting into the MLX resolver, but `mlx_sft.py` writes flat — so your exact symptom survived. You confirmed that independently and, more usefully, explained *why it was invisible*: your own reproduction config sets no `experiment_name`, so the bug could not fire on it. A test written from a reproduction alone would have kept missing it.

It is fixed structurally rather than conditionally — `_resolve_mlx_checkpoint` takes no `experiment_name` parameter at all — and the coupling between the writer's layout and the resolver's assumption is now pinned by a test that runs `train()` for real and resolves what it just wrote, in both directions.

**And your measurement retracted one of my two blocking findings.** I claimed `load_weights(strict=False)` might silently drop tensors; you ran it and posted numbers — identical trainable parameter count before and after, `lora_b` moving from exactly 0.0 to -0.003909. A freshly initialised `lora_b` is zero by construction, so a non-zero value cannot have come from anywhere but the file. That closed the question instead of continuing the argument, and it is credited in both the changelog fragment and the merge commit.

Verified before merging: 7 mutations, 7 killed, each by exactly one named test; zero deleted test lines; 14/14 CI including real Apple Silicon `mlx-smoke`.

**Two limits stated rather than hidden**, both now follow-ups:

1. No CI job exercises MLX `--resume` end to end on real mlx-lm — `mlx-smoke` trains but never resumes.
2. `strict=False` will still silently drop every tensor if a checkpoint was saved under a different `lora.r` or `target_modules`, and the run prints two green messages. The "declares at least one tensor" guard is the only mitigation reachable without mlx.

Try: `soup train --config soup.yaml --resume auto` with `backend: mlx`.

Reported and diagnosed by @imahsanali; fixed by @SID-6921.

**Comment by imahsanali** [comment]
Thanks for the clarification and for giving credit in the issue — I really appreciate it!

I’d like to take **#623**, the remaining half of **#366**, regarding the CPU-only assertion for `_build_source(..., pin=False, tier="ram")`.

I’ll work on this one and submit a PR once it’s ready. Thanks again for making the issue available!
